lib/mayaUsd/nodes/proxyShapeBase.cpp

```diff
@@ -768,6 +819,99 @@ UsdTimeCode MayaUsdProxyShapeBase::GetOutputTime(MDataBlock dataBlock) const
+    // ---------------------------------------------------------------------------------
+
+    // first, read the input value from "outStageDataAttr". outStageDataAttr gets computed when
+    // we get the stage on the proxy. If there is no incoming data, we return right away.
+    MDataHandle srcInDataCachedHandle = srcDataBlock.inputValue(outStageDataAttr, &retValue);
+    if (srcInDataCachedHandle.data().isNull()) {
+        return;
+    }
+
+    // query form the destination block to make sure inStageDataCachedAttr attribute is clean. If it
+    // is clean that means we already have the attr value.
+    MDataBlock dataBlock = forceCache();
+    if (dataBlock.isClean(inStageDataCachedAttr)) {
+        return;
+    }
+
+    // get the handle inDataCachedHandle and return if it doesn't have the data.
+    MDataHandle inDataCachedHandle = dataBlock.outputValue(inStageDataCachedAttr, &retValue);
+    if (inDataCachedHandle.data().isNull()) {
+        return;
+    }
+
+    MayaUsdStageData* srcInData
+        = dynamic_cast<MayaUsdStageData*>(srcInDataCachedHandle.asPluginData());
+    if (!srcInData || !srcInData->stage) {
+        return;
+    }
+
+    // get the pointer to source stage
+    UsdStageRefPtr srcUsdStage = srcInData->stage;
+
+    // transfer session layer
+    // session layer is never shared so transfer its content always.
+    SdfLayerRefPtr sessionLayer
+        = SdfLayer::CreateAnonymous(kAnonymousLayerName + kSessionLayerPostfix + ".usda");
+    sessionLayer->TransferContent(srcUsdStage->GetSessionLayer());
+
+    // decide if the root layer needs to be shared or deep copied.
+    SdfLayerRefPtr rootLayer = SdfLayer::CreateAnonymous(kAnonymousLayerName);
+    if (srcUsdStage->GetRootLayer()->IsAnonymous()) {
+        rootLayer->TransferContent(srcUsdStage->GetRootLayer());
+    } else {
+        rootLayer = srcUsdStage->GetRootLayer();
+    }
+
+    // create a new usd stage from the root and session layers
+    UsdStageRefPtr newUsdStage
+        = UsdStage::OpenMasked(rootLayer, sessionLayer, UsdStagePopulationMask::All());
+    TF_VERIFY(newUsdStage);
+
+    // handle edit target for session and root layers.
+    // setting edit target for SubLayers is handled separately.
+    auto srcCurrentTargetLayer = srcUsdStage->GetEditTarget().GetLayer();
+    auto isSessionLayer
+        = srcCurrentTargetLayer->GetIdentifier().find(kSessionLayerPostfix) != std::string::npos;
+    auto isAnonymous = srcUsdStage->GetRootLayer()->IsAnonymous();
+    if (isSessionLayer) {
+        newUsdStage->SetEditTarget(newUsdStage->GetSessionLayer());
+    } else if (!isSessionLayer && !isAnonymous) {
+        newUsdStage->SetEditTarget(srcCurrentTargetLayer);
+    }
+
+    // recursively create new anon Sublayer(s) for session and root layers
+    createNewAnonSubLayerRecursive(newUsdStage, srcCurrentTargetLayer, sessionLayer);
+    createNewAnonSubLayerRecursive(newUsdStage, srcCurrentTargetLayer, rootLayer);
+
```

Comment:
We could have nested Sublayer hierarchy under both "Session" and "Root" layers.

We recursively need to create a deep copy Sublayer(s) if the parent layer is AnonymousLayer.